contracts/roulette/lib.rs

```diff
@@ -0,0 +1,388 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+/// The Roulette
+/// - there is a window of length N blocks for users to place their bets
+/// - there are M bets allowed in each such block
+/// - after that no more bets can be placed until spin is called and the winnings are paid out
+#[ink::contract]
+mod roulette {
+
+    #[cfg(feature = "std")]
+    use ink::storage::traits::StorageLayout;
+    use ink::{
+        codegen::EmitEvent,
+        env::{
+            call::{build_call, ExecutionInput},
+            set_code_hash, DefaultEnvironment, Error as InkEnvError,
+        },
+        prelude::{format, string::String},
+        reflect::ContractEventBase,
+        storage::{traits::ManualKey, Lazy, Mapping},
+    };
+    use scale::{Decode, Encode};
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum RouletteError {
+        InkEnvError(String),
+        ArithmethicError,
+        BetAmountIsTooSmall,
+        PlayerAlreadyPlacedABet,
+        NoMoreBetsCanBeMade,
+        BettingPeriodNotOver,
+        NativeTransferFailed(String),
+        NotEnoughBalance,
+        CallerIsNotTheHouseOwner,
+    }
+
+    impl From<InkEnvError> for RouletteError {
+        fn from(e: InkEnvError) -> Self {
+            RouletteError::InkEnvError(format!("{e:?}"))
+        }
+    }
+
+    pub type Selector = [u8; 4];
+    pub type Result<T> = core::result::Result<T, RouletteError>;
+    pub type Event = <Roulette as ContractEventBase>::Type;
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct BetPlaced {
+        #[ink(topic)]
+        player: AccountId,
+        #[ink(topic)]
+        bet_type: BetType,
+        amount: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct WheelSpin {
+        winning_number: u8,
+    }
+
+    #[derive(Debug, Encode, Decode, Clone, Copy, PartialEq, Eq)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
+    pub enum BetType {
+        Number(u8),
+        Red,   // even
+        Black, // red
+    }
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
+    pub struct Bet {
+        pub player: AccountId,
+        pub bet_type: BetType,
+        pub amount: Balance,
+    }
+
+    #[derive(Debug)]
+    #[ink::storage_item]
+    pub struct Data {
+        /// represents the contract owner, defaults to the initializer of the contract
+        pub house: AccountId,
+        /// How long does the betting period last? (measured in blocks)
+        pub betting_period_length: BlockNumber,
+        /// When did this betting period start? (measured in blocks)
+        pub betting_period_start: BlockNumber,
```

Comment:
This is interesting. I did not have any such periods. I just let anyone place bets any time. There is no risk of getting your bet locked up for a long time because anyone can call `spin` at any time.

That being said, I also didn't bound my bet storage.

Do you think either approach is objectively better? Or are they just different?

Reply:
A fixed betting period is definitely more like a real-world roulette table.

Reply:
if you don't bound the storage and then loop over it, chances are it will never fit into one block